src/java.base/share/classes/java/text/ChoiceFormat.java

```diff
@@ -249,86 +251,110 @@ public void applyPattern(String newPattern) {
+                        }
+                        Segment.LIMIT.bldr.setLength(0);
+                        seg = Segment.FORMAT;
+                    }
+                    break;
+                case '|':
+                    if (inQuote) {
+                        seg.bldr.append(ch);
+                    } else {
+                        if (seg != Segment.FORMAT) {
+                            // Discard incorrect portion and finish building cFmt
+                            break;
+                        }
+                        // Insert an entry into the format and limit arrays
+                        // and switch to parsing limit
+                        limits.add(limit);
+                        formats.add(Segment.FORMAT.bldr.toString());
+                        Segment.FORMAT.bldr.setLength(0);
+                        seg = Segment.LIMIT;
+                    }
+                    break;
+                default:
+                    seg.bldr.append(ch);
+            }
+        }
 
-                String tempBuffer = segments[0].toString();
-                if (tempBuffer.equals("\u221E")) {
-                    startValue = Double.POSITIVE_INFINITY;
-                } else if (tempBuffer.equals("-\u221E")) {
-                    startValue = Double.NEGATIVE_INFINITY;
-                } else {
-                    startValue = Double.parseDouble(tempBuffer);
-                }
+        // clean up last one (SubPattern without trailing '|')
+        if (seg == Segment.FORMAT) {
+            limits.add(limit);
+            formats.add(Segment.FORMAT.bldr.toString());
+        }
+        choiceLimits = limits.stream().mapToDouble(d -> d).toArray();
+        choiceFormats = formats.toArray(new String[0]);
+    }
 
-                if (ch == '<' && startValue != Double.POSITIVE_INFINITY &&
-                        startValue != Double.NEGATIVE_INFINITY) {
-                    startValue = nextDouble(startValue);
-                }
-                if (startValue <= oldStartValue) {
-                    throw new IllegalArgumentException("Incorrect order of"
-                            + " intervals, must be in ascending order");
-                }
-                segments[0].setLength(0);
-                part = 1;
-            } else if (ch == '|') {
-                if (count == newChoiceLimits.length) {
-                    newChoiceLimits = doubleArraySize(newChoiceLimits);
-                    newChoiceFormats = doubleArraySize(newChoiceFormats);
-                }
-                newChoiceLimits[count] = startValue;
-                newChoiceFormats[count] = segments[1].toString();
-                ++count;
-                oldStartValue = startValue;
-                segments[1].setLength(0);
-                part = 0;
-            } else {
-                segments[part].append(ch);
-            }
+    // Used to explicitly define the segment mode while applying a pattern
+    private enum Segment {
```

Comment:
Do we need a new enum? Would introducing a new enum solve something that cannot be done with the existing implementation?

Reply:
No, we don't _need_ a new enum, although I prefer it for readability.

For example, I think the following is more clear:
- `Segment.LIMIT.bldr.toString()`over `segments[1].toString();`.
- `seg = Segment.LIMIT;` over `part = 0;`

Can also no longer do something such as `part = 2;` even if unlikely.

Although I can see the argument of not wanting an enum as `part` is either only `0` or `1`. Either is OK with me.

Reply:
OK, now the new enum can be shared with multiple threads with this implementation, but I guess it is OK as ChoiceFormat does not guarantee synchronization.